middleware/reverse/reverse.go

```diff
@@ -0,0 +1,109 @@
+package reverse
+
+import (
+	"net"
+
+	"github.com/miekg/coredns/middleware"
+	"golang.org/x/net/context"
+	"github.com/miekg/dns"
+	"github.com/miekg/coredns/request"
+	"github.com/miekg/coredns/middleware/pkg/dnsutil"
+)
+
+// Reverse provides dynamic reverse dns and the related forward rr
+type Reverse struct {
+	Next     middleware.Handler
+	Networks Networks
+}
+
+// ServeDNS implements the middleware.Handler interface.
+func (reverse Reverse) ServeDNS(ctx context.Context, w dns.ResponseWriter, r *dns.Msg) (int, error) {
+	var rr dns.RR
+
+	state := request.Request{W: w, Req: r}
+	m := new(dns.Msg)
+	m.SetReply(r)
+	m.Authoritative, m.RecursionAvailable, m.Compress = true, true, true
+
+	// Reverse Lookup
+	if state.QType() == dns.TypePTR {
+
+		address := dnsutil.ExtractAddressFromReverse(state.Name())
+		if address == "" {
+			// Invalid PTR request
+			return dns.RcodeFormatError, nil
+		}
+
+		ip := net.ParseIP(address)
+		// loop through the configured networks
+		for _, n := range reverse.Networks {
+			if (n.Ipnet.Contains(ip)) {
+				rr = &dns.PTR{
+					Hdr:  dns.RR_Header{Name: state.QName(), Rrtype: dns.TypePTR, Class: dns.ClassINET, Ttl: n.Ttl},
+					Ptr: n.ipToHostname(&ip),
+				}
+				break
+			}
+		}
+
+	} else {
+
+		// Forward Lookup
+		switch state.QType(){
+		case dns.TypeA:
+			for _, n := range reverse.Networks {
+				// skip if requesting an v4 address and network is not v4
+				if n.Ipnet.IP.To4() == nil {
+					continue
+				}
+
+				if dns.IsSubDomain(n.Domain, state.Name()) {
+					result := n.hostnameToIp(state.Name())
+					if result != nil {
+						rr = &dns.A{
+							Hdr:  dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeA, Class: dns.ClassINET, Ttl: n.Ttl},
+							A: result,
+						}
+						break
+					}
+				}
+			}
+
+		case dns.TypeAAAA:
+			for _, n := range reverse.Networks {
+				// Do not use To16 which tries to make v4 in v6
+				if n.Ipnet.IP.To4() != nil {
+					continue
+				}
+
+				if dns.IsSubDomain(n.Domain, state.Name()) {
+					result := n.hostnameToIp(state.Name())
+					if result != nil {
+						rr = &dns.AAAA{
+							Hdr:  dns.RR_Header{Name: state.QName(), Rrtype: dns.TypeAAAA, Class: dns.ClassINET, Ttl: n.Ttl},
+							AAAA: result,
+						}
+						break
+					}
+				}
+			}
+		}
+	}
+
+	if rr == nil {
+		if reverse.Next == nil {
+			return dns.RcodeNameError, nil
+		}
+		return reverse.Next.ServeDNS(ctx, w, r)
```

Comment:
we need to think about what this mean. Had a chat with @johnbelamaric about chaining middleware like this. We should probably make it explicit that we call another middleware. I can see the power (because now you don't care about SOA/NS whatever)

Seeing this use case I think we should allow it.